File: crates/tbdex/src/messages/order_status.rs
use super::{MessageKind, MessageMetadata, Result};
use crate::{
    json::{FromJson, ToJson},
    json_schemas::generated::{MESSAGE_JSON_SCHEMA, ORDER_STATUS_DATA_JSON_SCHEMA},
    DEFAULT_PROTOCOL_VERSION,
};
use chrono::Utc;
use serde::{Deserialize, Serialize};
use web5::dids::bearer_did::BearerDid;

/// Represents an Order Status message in the tbDEX protocol.
///
/// An Order Status message is sent from a PFI to Alice to communicate
/// the current status of an ongoing order or exchange process.
#[derive(Clone, Serialize, Deserialize, Debug, PartialEq)]
pub struct OrderStatus {
    /// Metadata about the message, including sender, recipient, and protocol information.
    pub metadata: MessageMetadata,

    /// The public data part of the Order Status, such as the current status of the order.
    pub data: OrderStatusData,

    /// The signature verifying the authenticity and integrity of the Order Status message.
    pub signature: String,
}

impl ToJson for OrderStatus {}
impl FromJson for OrderStatus {}

impl OrderStatus {
    /// Creates a new Order Status message.
    ///
    /// # Arguments
    ///
    /// * `to` - The DID of the recipient (Alice).
    /// * `from` - The DID of the sender (the PFI).
    /// * `exchange_id` - The exchange ID shared between Alice and the PFI.
    /// * `data` - The data containing the current status of the order.
    /// * `protocol` - Optional protocol version; defaults to the current version if not provided.
    /// * `external_id` - Optional external ID for additional identification.
    ///
    /// # Returns
    ///
    /// A new instance of `OrderStatus` containing the metadata, data, and an empty signature.
    pub fn create(
        to: &str,
        from: &str,
        exchange_id: &str,
        data: &OrderStatusData,
        protocol: Option<String>,
        external_id: Option<String>,
    ) -> Result<Self> {
        let metadata = MessageMetadata {
            from: from.to_string(),
            to: to.to_string(),
            kind: MessageKind::OrderStatus,
            id: MessageKind::OrderStatus.typesafe_id()?,
            exchange_id: exchange_id.to_string(),
            external_id,
            protocol: protocol.unwrap_or_else(|| DEFAULT_PROTOCOL_VERSION.to_string()),
            created_at: Utc::now().to_rfc3339(),
        };

        let order_status = Self {
            metadata: metadata.clone(),
            data: data.clone(),
            signature: String::default(),
        };

        Ok(order_status)
    }

    /// Signs the Order Status message using the provided Bearer DID.
    ///
    /// # Arguments
    ///
    /// * `bearer_did` - The DID to sign the Order Status message.
    ///
    /// # Returns
    ///
    /// An empty result, or an error if the signing process fails.
    pub fn sign(&mut self, bearer_did: &BearerDid) -> Result<()> {
        self.signature = crate::signature::sign(
            bearer_did,
            &serde_json::to_value(&self.metadata)?,
            &serde_json::to_value(&self.data)?,
        )?;
        Ok(())
    }

    /// Verifies the validity of the Order Status message.
    ///
    /// This method ensures that the message adheres to its JSON schema
    /// and verifies the signature to ensure authenticity and integrity.
    ///
    /// # Returns
    ///
    /// An empty result if verification succeeds, or an error if verification fails.
    pub fn verify(&self) -> Result<()> {
        // verify resource json schema
        crate::json_schemas::validate_from_str(MESSAGE_JSON_SCHEMA, self)?;

        // verify data json schema
        crate::json_schemas::validate_from_str(ORDER_STATUS_DATA_JSON_SCHEMA, &self.data)?;

        // verify signature
        crate::signature::verify(
            &self.metadata.from,
            &serde_json::to_value(self.metadata.clone())?,
            &serde_json::to_value(self.data.clone())?,
            &self.signature,
        )?;

        Ok(())
    }
}

/// Represents the data for an Order Status message in the tbDEX protocol.
///
/// This includes the current status of the order and any optional details
/// providing additional information about the status.
#[derive(Clone, Serialize, Deserialize, Debug, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct OrderStatusData {
    /// The current status of the order (e.g., PayinPending, PayinSettled).
    pub status: Status,

    /// Optional additional details about the current status.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub details: Option<String>,
}

/// Represents the possible statuses for an order in the tbDEX protocol.
///
/// Each status indicates a specific stage in the lifecycle of an order.
#[derive(Debug, Default, Deserialize, PartialEq, Serialize, Clone)]
#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
pub enum Status {
    /// Indicates that the PFI is awaiting payment from Alice.
    #[default]
    PayinPending,

    /// Indicates that payment from Alice has been initiated.
    PayinInitiated,

    /// Indicates that payment from Alice has been successfully settled into the PFI's account.
    PayinSettled,

    /// Indicates that payment from Alice has failed.
    PayinFailed,

    /// Indicates that payment from Alice was not received before the quote expired.
    PayinExpired,

    /// Indicates that the payout to Alice is pending further processing.
    PayoutPending,

    /// Indicates that the payout to Alice has been initiated.
    PayoutInitiated,

    /// Indicates that the payout to Alice has been successfully settled.
    PayoutSettled,

    /// Indicates that the payout to Alice has failed.
    PayoutFailed,

    /// Indicates that a refund of Alice's payin is pending further processing.
    RefundPending,

    /// Indicates that the refund of Alice's payin has been initiated.
    RefundInitiated,

    /// Indicates that the refund of Alice's payin has been successfully settled.
    RefundSettled,

    /// Indicates that the refund of Alice's payin has failed.
    RefundFailed,
}

#[cfg(test)]
mod tbdex_test_vectors_protocol {
    use super::*;
    use std::fs;

    #[derive(Debug, serde::Deserialize)]
    pub struct TestVector {
        pub input: String,
        pub output: OrderStatus,
    }

    #[test]
    fn parse_orderstatus() {
        let path = "../../tbdex/hosted/test-vectors/protocol/vectors/parse-orderstatus.json";
        let test_vector_json: String = fs::read_to_string(path).unwrap();

        let test_vector: TestVector = serde_json::from_str(&test_vector_json).unwrap();
        let parsed_order_status: OrderStatus =
            OrderStatus::from_json_string(&test_vector.input).unwrap();

        assert_eq!(test_vector.output, parsed_order_status);
    }
}

File: bound/kt/src/main/kotlin/tbdex/sdk/messages/OrderStatus.kt
package tbdex.sdk.messages

import tbdex.sdk.TbdexException
import tbdex.sdk.http.ReplyToMessage
import tbdex.sdk.rust.OrderStatusStatus as RustCoreStatus
import tbdex.sdk.rust.OrderStatus as RustCoreOrderStatus
import tbdex.sdk.rust.OrderStatusDataData as RustCoreOrderStatusData
import tbdex.sdk.rust.fromWeb5
import tbdex.sdk.rust.BearerDid as RustCoreBearerDid
import web5.sdk.dids.BearerDid

/**
 * Represents the possible statuses for an order in the tbDEX protocol.
 *
 * Each status indicates a specific stage in the lifecycle of an order.
 */
enum class Status {
    PAYIN_PENDING,
    PAYIN_INITIATED,
    PAYIN_SETTLED,
    PAYIN_FAILED,
    PAYIN_EXPIRED,
    PAYOUT_PENDING,
    PAYOUT_INITIATED,
    PAYOUT_SETTLED,
    PAYOUT_FAILED,
    REFUND_PENDING,
    REFUND_INITIATED,
    REFUND_SETTLED,
    REFUND_FAILED;

    companion object {
        internal fun fromRustCore(rustCore: RustCoreStatus): Status {
            return when (rustCore) {
                RustCoreStatus.PAYIN_PENDING -> PAYIN_PENDING
                RustCoreStatus.PAYIN_INITIATED -> PAYIN_INITIATED
                RustCoreStatus.PAYIN_SETTLED -> PAYIN_SETTLED
                RustCoreStatus.PAYIN_FAILED -> PAYIN_FAILED
                RustCoreStatus.PAYIN_EXPIRED -> PAYIN_EXPIRED
                RustCoreStatus.PAYOUT_PENDING -> PAYOUT_PENDING
                RustCoreStatus.PAYOUT_INITIATED -> PAYOUT_INITIATED
                RustCoreStatus.PAYOUT_SETTLED -> PAYOUT_SETTLED
                RustCoreStatus.PAYOUT_FAILED -> PAYOUT_FAILED
                RustCoreStatus.REFUND_PENDING -> REFUND_PENDING
                RustCoreStatus.REFUND_INITIATED -> REFUND_INITIATED
                RustCoreStatus.REFUND_SETTLED -> REFUND_SETTLED
                RustCoreStatus.REFUND_FAILED -> REFUND_FAILED
            }
        }
    }

    internal fun toRustCore(): RustCoreStatus {
        return when (this) {
            PAYIN_PENDING -> RustCoreStatus.PAYIN_PENDING
            PAYIN_INITIATED -> RustCoreStatus.PAYIN_INITIATED
            PAYIN_SETTLED -> RustCoreStatus.PAYIN_SETTLED
            PAYIN_FAILED -> RustCoreStatus.PAYIN_FAILED
            PAYIN_EXPIRED -> RustCoreStatus.PAYIN_EXPIRED
            PAYOUT_PENDING -> RustCoreStatus.PAYOUT_PENDING
            PAYOUT_INITIATED -> RustCoreStatus.PAYOUT_INITIATED
            PAYOUT_SETTLED -> RustCoreStatus.PAYOUT_SETTLED
            PAYOUT_FAILED -> RustCoreStatus.PAYOUT_FAILED
            REFUND_PENDING -> RustCoreStatus.REFUND_PENDING
            REFUND_INITIATED -> RustCoreStatus.REFUND_INITIATED
            REFUND_SETTLED -> RustCoreStatus.REFUND_SETTLED
            REFUND_FAILED -> RustCoreStatus.REFUND_FAILED
        }
    }
}

/**
 * Represents the data of an order status in the tbDEX protocol.
 *
 * @property status The current status of the order (e.g., PAYIN_PENDING, PAYIN_SETTLED).
 * @property details Optional details providing additional information about the status.
 */
data class OrderStatusData (
    val status: Status,
    val details: String?
) {
    companion object {
        internal fun fromRustCore(rustCore: RustCoreOrderStatusData): OrderStatusData {
            return OrderStatusData(Status.fromRustCore(rustCore.status), rustCore.details)
        }
    }

    internal fun toRustCore(): RustCoreOrderStatusData {
        return RustCoreOrderStatusData(status.toRustCore(), details)
    }
}

/**
 * Represents an Order Status message in the tbDEX protocol.
 *
 * An Order Status message is sent by a PFI to Alice to communicate
 * the current status of an ongoing order or exchange process.
 *
 * @property metadata Metadata about the message, including sender, recipient, and protocol information.
 * @property data The data part of the Order Status, such as the current status of the order.
 * @property signature The signature verifying the authenticity and integrity of the Order Status message.
 * @property rustCoreOrderStatus The underlying RustCore representation of the Order Status.
 */
data class OrderStatus private constructor(
    val metadata: MessageMetadata,
    val data: OrderStatusData,
    val signature: String,
    internal val rustCoreOrderStatus: RustCoreOrderStatus
): Message, ReplyToMessage {
    companion object {
        /**
         * Creates a new Order Status message.
         *
         * @param to The DID of the recipient (Alice).
         * @param from The DID of the sender (the PFI).
         * @param exchangeId The exchange ID shared between Alice and the PFI.
         * @param data The data containing the order status.
         * @param protocol Optional protocol version.
         * @param externalId Optional external identifier.
         * @return The newly created Order Status message.
         * @throws TbdexException if the creation process fails.
         */
        fun create(
            to: String,
            from: String,
            exchangeId: String,
            data: OrderStatusData,
            protocol: String? = null,
            externalId: String? = null
        ): OrderStatus {
            try {
                val rustCoreOrderStatus = RustCoreOrderStatus.create(to, from, exchangeId, data.toRustCore(), protocol, externalId)
                val rustCoreData = rustCoreOrderStatus.getData()
                return OrderStatus(
                    MessageMetadata.fromRustCore(rustCoreData.metadata),
                    OrderStatusData.fromRustCore(rustCoreData.data),
                    rustCoreData.signature,
                    rustCoreOrderStatus
                )
            } catch (e: tbdex.sdk.rust.TbdexException.Exception) {
                throw TbdexException.fromRustCore(e)
            }
        }

        /**
         * Parses an Order Status from a JSON string.
         *
         * @param json The JSON string representing the Order Status.
         * @return The deserialized Order Status message.
         * @throws TbdexException if parsing fails.
         */
        fun fromJsonString(json: String): OrderStatus {
            try {
                val rustCoreOrderStatus = RustCoreOrderStatus.fromJsonString(json)
                val rustCoreData = rustCoreOrderStatus.getData()
                return OrderStatus(
                    MessageMetadata.fromRustCore(rustCoreData.metadata),
                    OrderStatusData.fromRustCore(rustCoreData.data),
                    rustCoreData.signature,
                    rustCoreOrderStatus
                )
            } catch (e: tbdex.sdk.rust.TbdexException.Exception) {
                throw TbdexException.fromRustCore(e)
            }
        }

        internal fun fromRustCoreOrderStatus(rustCoreOrderStatus: RustCoreOrderStatus): OrderStatus {
            val rustCoreData = rustCoreOrderStatus.getData()
            return OrderStatus(
                MessageMetadata.fromRustCore(rustCoreData.metadata),
                OrderStatusData.fromRustCore(rustCoreData.data),
                rustCoreData.signature,
                rustCoreOrderStatus
            )
        }
    }

    /**
     * Serializes the Order Status to a JSON string.
     *
     * @return The serialized JSON string of the Order Status.
     * @throws TbdexException if serialization fails.
     */
    fun toJsonString(): String {
        try {
            return rustCoreOrderStatus.toJsonString()
        } catch (e: tbdex.sdk.rust.TbdexException.Exception) {
            throw TbdexException.fromRustCore(e)
        }
    }

    /**
     * Signs the Order Status using the provided Bearer DID.
     *
     * @param bearerDid The Bearer DID used to sign the Order Status.
     * @throws TbdexException if the signing process fails.
     */
    fun sign(bearerDid: BearerDid) {
        try {
            rustCoreOrderStatus.sign(RustCoreBearerDid.fromWeb5(bearerDid))
        } catch (e: tbdex.sdk.rust.TbdexException.Exception) {
            throw TbdexException.fromRustCore(e)
        }
    }

    /**
     * Verifies the Order Status's signature and validity.
     *
     * @throws TbdexException if verification fails.
     */
    fun verify() {
        try {
            rustCoreOrderStatus.verify()
        } catch (e: tbdex.sdk.rust.TbdexException.Exception) {
            throw TbdexException.fromRustCore(e)
        }
    }
}